File: algorithm/src/leetcode/MaximumSubarray_53.java
package leetcode;

// #53 - Maximum Subarray

public class MaximumSubarray_53 {
	public static void main(String[] args) {
		int[] nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
		System.out.println("maxSubArray(nums) = " + maxSubArray(nums));
	}

	public static int maxSubArray(int[] nums) {
		if(nums.length == 1) return nums[0];

		int max = nums[0], sum = 0;
		for (int num : nums) {
			sum += num;
			if(sum > max) {
				max = sum;
			}
			if(sum < 0) sum = 0;
		}
		return max;
	}
}

File: ruby/problem_solving/leetcode/MaximumSubarray_53.rb


# @param {Integer[]} nums
# @return {Integer}

def max_sub_array(nums)
  sum = max = nums.first
  nums[1..].each do |num|
    sum = if sum.negative?
            num
          else
            sum + num
          end
    max = [sum, max].max
  end
  max
end

# nums = [1]
nums = [-2,1,-3,4,-1,2,1,-5,4]
p max_sub_array(nums)
